pkg/migration/migration.go

```diff
@@ -0,0 +1,145 @@
+	// TODO
+	/*
+		add validation to check source,destination storageclass and Rook namespace
+	*/
+
+	// Create Kubernetes Client
+	logger.DefaultLog("Create Kubernetes Client")
+	client, err := k8sutil.NewClient(kubeConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create kubernetes client: %v", err)
+	}
+
+	logger.DefaultLog("List all the PVC from the source storageclass")
+	pvcs, err := k8sutil.ListAllPVCWithStorageclass(client, sourceStorageClass)
+	if err != nil {
+		return fmt.Errorf("failed to list PVCs from the storageclass: %v", err)
+	}
+	if pvcs == nil || len(*pvcs) == 0 {
+		logger.DefaultLog("no PVCs found with storageclass: %v", sourceStorageClass)
+		return nil
+	}
+	logger.DefaultLog("%d PVCs found with sourceStorageClass %s ", len(*pvcs), sourceStorageClass)
+
+	logger.DefaultLog("Start Migration of PVCs to CSI")
+	for _, pvc := range *pvcs {
+		err = migratePVC(client, pvc, destinationStorageClass, rookNamespace, cephClusterNamespace)
+		if err != nil {
+			return fmt.Errorf("failed to migrate PVC %s : %v", pvc.Name, err)
+		}
+	}
+	logger.DefaultLog("Successfully migrated all the PVCs to CSI")
+
+	return err
+}
+
+// migratePVC migrates a PVC to CSI.
+func migratePVC(client *k8s.Clientset, pvc v1.PersistentVolumeClaim, destinationStorageClass,
+	rookNamespace, cephClusterNamespace string) error {
+
+	logger.DefaultLog("migrating PVC %q", pvc.Name)
+
+	logger.DefaultLog("Fetch PV information from PVC %s", pvc.Name)
+	pv, err := k8sutil.GetPV(client, pvc.Spec.VolumeName)
+	if err != nil {
+		return fmt.Errorf("failed to get PV object with name %s: %v", pvc.Spec.VolumeName, err)
+	}
+	logger.DefaultLog("PV found %q ", pv.Name)
+
+	logger.DefaultLog("Update Reclaim policy from Delete to Reclaim for PV: %s", pv.Name)
+	err = k8sutil.UpdateReclaimPolicy(client, pv)
+	if err != nil {
+		return fmt.Errorf("failed to update ReclaimPolicy for PV object %s: %v", pvc.Spec.VolumeName, err)
+	}
+
+	logger.DefaultLog("Retrieving old ceph volume name from PV object: %s", pv.Name)
+	rbdImageName := k8sutil.GetVolumeName(pv)
+	if rbdImageName == "" {
+		return fmt.Errorf("rbdImageName cannot be empty in Flex PV object: %v", pv)
+	}
+	logger.DefaultLog("Flex rbdImageName name %v ", rbdImageName)
+
+	logger.DefaultLog("Deleting pvc object: %s", pvc.Name)
+	err = k8sutil.DeletePVC(client, &pvc) // nolint:gosec // skip gosec as pvc is accessed via it's reference.
+	if err != nil {
+		return fmt.Errorf("failed to Delete PVC object %s: %v", pvc.Name, err)
+	}
+
+	logger.DefaultLog("Generate new PVC with same name in destination storageclass")
+	csiPVC := k8sutil.GenerateCSIPVC(destinationStorageClass, &pvc) // nolint:gosec // skip gosec as pvc is accessed via it's reference.
+	logger.DefaultLog("Structure of the generated PVC in destination Storageclass %v ", csiPVC)
+
+	logger.DefaultLog("Create new csi pvc")
+	csiPV, err := k8sutil.CreatePVC(client, csiPVC, pvcCreateTimeout)
```

Comment:
This is a PVC, not PV, right?
```suggestion
	csiPVC, err := k8sutil.CreatePVC(client, csiPVC, pvcCreateTimeout)
```

Reply:
Yes, the function creates the PVC, but returns the pv object of the created PVC

Reply:
Ok I see now that it waits for the PV to be created, then returns it. So why do we also need to wait for PV creation on line 97 below? We already have the PV. That's actually waiting for the rbd image creation, right? My earlier suggestion on the method name wasn't quite right.

Reply:
Yes correct, the `WaitForPVCreation` function waits for rbd image creation, so that the imagename is accessible from the PV spec

Reply:
How about renaming it to `WaitForRBDImage()`?